### PR TITLE
Last nits for v1.7 release

### PR DESCRIPTION
**Pull request by pv3nturi**
Last nits for v1.7 release:
- Added encoding for uri with [ ] chars
- Removed no more used dynamic config properties
- Disabled auto-redirect of proxy clients to endpoints
- Added compression on requests (request.compression.enabled; default true) and responses (response.compression.threshold; default 0 bytes = always enabled; -1 to disable the compression)
- Added graceful shutdown for connection pools: new property disposetimeout (default = connectionsmanager.disposetimeout)
- Added Connection Pools page in Carapace Admin UI
- Fixed maxconnectionsperendpoint limit for connection pools

### CODE REVIEW

File: carapace-server/src/main/java/org/carapaceproxy/core/RuntimeServerConfiguration.java [hunk 1]
```diff
@@ -90,7 +90,6 @@
     private Set<String> domainsCheckerIPAddresses;
     private List<String> supportedSSLProtocols = null;
     private int ocspStaplingManagerPeriod = 0;
-    private boolean requestsHeaderDebugEnabled = false;
```
Comment (pv3nturi) [inline carapace-server/src/main/java/org/carapaceproxy/core/RuntimeServerConfiguration.java:93]: unused

File: carapace-server/src/main/java/org/carapaceproxy/core/Listeners.java [hunk 2]
```diff
@@ -241,6 +243,19 @@ protected SslHandler newSslHandler(SslContext context, ByteBufAllocator allocato
                         };
                         channel.pipeline().addFirst(sni);
                     }
+                    channel.pipeline().addAfter(NettyPipeline.HttpCodec, "uriEncoder", new ChannelInboundHandlerAdapter() {
+                        @Override
+                        public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                            if (msg instanceof HttpRequest) {
+                                HttpRequest request = (HttpRequest) msg;
+                                request.setUri(request.uri()
+                                        .replaceAll("\\[", "%5B")
```
Comment (pv3nturi) [inline carapace-server/src/main/java/org/carapaceproxy/core/Listeners.java:252]: Reactor Netty triggers an exception for not compliant URIs

File: carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java [hunk 3]
```diff
@@ -317,15 +318,23 @@ private RequestForwarder(ProxyRequest request, ContentsCache.ContentReceiver cac
...
+                if (stats != null) {
+                    CarapaceLogger.debug("Connection {0} stats: {1}", connectionConfig.getId(), stats.get(connectionConfig.getId()));
+                }
+                CarapaceLogger.debug("Max connections for {0}: {1}", connectionConfig.getId(), connectionProvider.maxConnectionsPerHost());
+            }
 
             client = HttpClient.create(connectionProvider)
                     .host(endpointHost)
                     .port(endpointPort)
+                    .followRedirect(false) // clients has to request the redirect, not the proxy
```
Comment (diegosalvi) [inline carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java:332]: +1

File: carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java [hunk 4]
```diff
@@ -469,32 +478,29 @@ public void reloadConfiguration(RuntimeServerConfiguration newConfiguration, Col
...
-                        .maxIdleTime(Duration.ofMillis(connectionPool.getIdleTimeout()))
-                        .metrics(true);
+                        .disposeTimeout(Duration.ofMillis(connectionPool.getDisposeTimeout()));
 
                 // max connections per endpoint limit setup
                 newEndpoints.forEach(be -> {
-                    builder.forRemoteHost(new InetSocketAddress(
-                            be.getHost(), be.getPort()),
-                            spec -> spec.maxConnections(connectionPool.getMaxConnectionsPerEndpoint())
+                    CarapaceLogger.debug("Setup max connections per endpoint {0}:{1} = {2} for connectionpool {3}",
```
Comment (dmercuriali) [inline carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java:493]: we should wrap this with 
if (CarapaceLogger.isLoggingDebugEnabled())

File: carapace-server/src/main/resources/conf/server.dynamic.properties [hunk 5]
```diff
@@ -54,7 +54,8 @@ route.100.action=cache-if-possible
 default.action.notfound=not-found
 default.action.internalerror=internal-error
 
-
+# Custom X-Proxy-Path header
+mapper.debug=true
```
Comment (diegosalvi) [inline carapace-server/src/main/resources/conf/server.dynamic.properties:58]: Should it be debug by default?

File: carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java [hunk 6]
```diff
@@ -508,25 +514,26 @@ public void reloadConfiguration(RuntimeServerConfiguration newConfiguration, Col
...
-            String hostName = t.getRemoteAddress().getHostName();
-            return connectionPools.entrySet().stream()
+        public Map.Entry<ConnectionPoolConfiguration, ConnectionProvider> apply(ProxyRequest request) {
+            String hostName = request.getRequestHostname();
+            Map.Entry<ConnectionPoolConfiguration, ConnectionProvider> selectedPool = connectionPools.entrySet().stream()
                     .filter(e -> Pattern.matches(e.getKey().getDomain(), hostName))
                     .findFirst()
                     .orElse(defaultConnectionPool);
+
+            CarapaceLogger.debug("Using connection {0} for domain {1}", selectedPool.getKey().getId(), hostName);
```
Comment (dmercuriali) [inline carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java:534]: we should wrap this with
if (CarapaceLogger.isLoggingDebugEnabled())

**Review by diegosalvi — COMMENTED** [review]
Good work!

Just one comment and can you add a test for url with [ and ] too?

File: carapace-server/src/main/resources/conf/server.dynamic.properties (hunk 5 above)
Comment (pv3nturi) [reply carapace-server/src/main/resources/conf/server.dynamic.properties:58]: In production environment is enabled, it's is very useful even in dev/debug environment.

File: carapace-ui/src/main/webapp/src/components/ConnectionPools.vue [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    export default {
+        name: "ConnectionPools",
+        data() {
+            return {
+                connectionPools: []
+            };
+        },
+        created() {
+            doGet("/api/connectionpools", data => {
+                console.log(data)
```
Comment (dmercuriali) [inline carapace-ui/src/main/webapp/src/components/ConnectionPools.vue:21]: forgot this?

File: carapace-server/src/main/resources/conf/server.dynamic.properties (hunk 5 above)
Comment (dmercuriali) [inline carapace-server/src/main/resources/conf/server.dynamic.properties:58]: i think we should set this to false by default, even if this is just an example file

**Review by dmercuriali — CHANGES_REQUESTED** [review]

File: carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java (hunk 6 above)
Comment (pv3nturi) [reply carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java:534]: Even if not wrapped it's visible at FINE level whether debug is disabled

**Review by dmercuriali — APPROVED** [review]

**Pull request merged**